### ISSUE TITLE
[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: DioError [DioErrorType.response]: Http status error [500]

### ISSUE THREAD

**Issue by Vedsaga** [flutterchina/dio#1310]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | e.g. flutter / ios / android                          |    andriod  |
| Platform Version               | e.g.  1.5.0 / 12.0 / 9.0                              |     emulator |
| Dio Version                    | e.g.  2.1.0 / 1.0.17                                  | 4.0.1 |
| Android Studio / Xcode Version | e.g. Android Studio 3.3.2 / Xcode 10.2.1              |   VS Code latest Version   |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |   100   |
| Repro with our demo prj        | e.g. does it happen with our demo project?            |      |
| Demo project link              | e.g. link to a demo project that highlights the issue |      |

### Issue Description and Steps
I am making post request. And, all the required argument where correctly passed still I am getting this error.

```dart
  Future<Map<String, dynamic>?> newShipment(
      NewShipmentRequest request, String token) async {
    try {
      final Response response = await CommonProvider.dio.post(
        AppURLs.newShipment,
        data: request.toMap(),
        options: Options(
          headers: {
            "Authorization": "Token $token",
          },
        ),
      );
      return response.asMap();
    } on DioError catch (e) {
      print("printng----e ---------");
      print(e);
      print('printing ----- e.response -----');
      print(e.response);
      print('printing --- e.error----');
      print(e.error);
      print('printing ---- e.requestOptions.');
      print(e.requestOptions);
    }
  }
```
When I do post request then it is expecting map as return but, instead I am getting html in return below output is from terminal..

```I/flutter ( 8820): printng----e ---------
I/flutter ( 8820): DioError [DioErrorType.response]: Http status error [500]
I/flutter ( 8820): #0      DioMixin.assureDioError (package:dio/src/dio_mixin.dart:820:20)
I/flutter ( 8820): #1      DioMixin._dispatchRequest (package:dio/src/dio_mixin.dart:680:13)
I/flutter ( 8820): <asynchronous suspension>
I/flutter ( 8820): #2      DioMixin.fetch.<anonymous closure>.<anonymous closure> (package:dio/src/dio_mixin.dart)
I/flutter ( 8820): <asynchronous suspension>
I/flutter ( 8820): printing ----- e.response -----
I/flutter ( 8820): <!DOCTYPE html>
I/flutter ( 8820): <html lang="en">
I/flutter ( 8820): <head>
I/flutter ( 8820):   <meta http-equiv="content-type" content="text/html; charset=utf-8">
I/flutter ( 8820):   <meta name="robots" content="NONE,NOARCHIVE">
I/flutter ( 8820):   <title>RuntimeError
I/flutter ( 8820):           at /register_shipment</title>
I/flutter ( 8820):   <style type="text/css">
I/flutter ( 8820):     html * { padding:0; margin:0; }
I/flutter ( 8820):     body * { padding:10px 20px; }
I/flutter ( 8820):     body * * { padding:0; }
I/flutter ( 8820):     body { font:small sans-serif; background-color:#fff; color:#000; }
I/flutter ( 8820):     body>div { border-bottom:1px solid #ddd; }
I/flutter ( 8820):     h1 { font-weight:normal; }
I/flutter ( 8820):     h2 { margin-bottom:.8em; }
I/flutter ( 8820):     h3 { margin:1em 0 .5em 0; }
I/flutter ( 8820):     h4 { margin:0 0 .5em 0; font-weight: normal; }
I/flutter ( 8820):     code, pre { font-size: 100%; white-space: pre-wrap; }
I/flutter ( 8820):     table { border:1px solid #ccc; border-collapse: collapse; width:100%; background:white; }
I/flutter ( 8820):     tbody td, tbody th { vertical-align:top; padding:2px 3px; }
I/flutter ( 8820):     thead th {
I/flutter ( 8820):       padding:1px 6px 1px 3px; background:#fefefe; text-align:left;
I/flutter ( 8820):       font-weight:normal; font-size:11px; border:1px solid #ddd;
I/flutter ( 8820):     }
I/flutter ( 8820):     tbody th { width:12em; text-align:right; c
I/flutter ( 8820): printing --- e.error----
I/flutter ( 8820): Http status error [500]
I/flutter ( 8820): printing ---- e.requestOptions.
I/flutter ( 8820): Instance of 'RequestOptions'```

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

**Comment by markfili** [comment]
this is what I have for you to try:
```
_client.interceptors.add(InterceptorsWrapper(onError: (error, handler) {
      if (error != null) {
        handler.reject(error);
      }
    }));
```

**Comment by Vedsaga** [comment]
> ```
> _client.interceptors.add(InterceptorsWrapper(onError: (error, handler) {
>       if (error != null) {
>         handler.reject(error);
>       }
>     }));
> ```

Where will this interceptors should be ?

**Comment by markfili** [comment]
after you create an instance of dio, you can add interceptors to it;
```
var dioClient = Dio();
dioClient.interceptors.add(InterceptorsWrapper(onError: (error, handler) {
      if (error != null) {
        handler.reject(error);
      }
    }));
```

**Comment by Vedsaga** [comment]
> ```
> var dioClient = Dio();
> dioClient.interceptors.add(InterceptorsWrapper(onError: (error, handler) {
>       if (error != null) {
>         handler.reject(error);
>       }
>     }));
> ```
Just 2 quick question because I am not much familiar with network part.... so, pls pardon me if I am wrong here, 

- When you say  `create an instance of dio` are you referring to  mean request call which I am making using dio? 
- What is the reason which is causing my post request to crash?

**Comment by markfili** [comment]
@Vedsaga I'm sorry, I see now I should've read the log you've pasted in the question.
Disregard what I've said, you don't seem to need the `handler.reject` or the interceptors, you need to check your backend server and see why it got a `RuntimeError` when calling its `register_shipment` endpoint. Your error is most probably server side and not in the app.

**Comment by Vedsaga** [comment]
> @Vedsaga I'm sorry, I see now I should've read the log you've pasted in the question. Disregard what I've said, you don't seem to need the `handler.reject` or the interceptors, you need to check your backend server and see why it got a `RuntimeError` when calling its `register_shipment` endpoint. Your error is most probably server side and not in the app.

Please don't say sorry, I really appreciate the time you took to go through my issue... I will check with backend team... Thanks you again for trying to me :)

**Comment by Vedsaga** [comment]
![image](https://user-images.githubusercontent.com/59083528/140329766-bd6b9c5c-5775-47fe-ad19-e874e2a889ff.png)
I found the issue, it was from app side actually API was crashing... because I was sending extra field which I was not supposed to send.... I should close this issue...


Really sorry for the trouble I caused...
